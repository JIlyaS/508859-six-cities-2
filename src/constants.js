export const RATING_PERSENT = 100;
export const MAX_RATING_COUNT = 5;
export const ICON = {
  url: `img/pin.svg`,
  size: [30, 30]
};
export const LEAFLET_OPTION = {
  center: [52.38333, 4.9],
  zoom: 12
};
export const TILE_LAYER = {
  img: `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`,
  attributions: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
};
