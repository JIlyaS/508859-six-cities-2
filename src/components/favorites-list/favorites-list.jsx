import React, {Fragment, PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';

import PageLayout from '../page-layout/page-layout';
import Operation from '../../operation/operation';
import OfferCard from '../offer-card/offer-card';
import FavoritesEmpty from '../favorites-empty/favorites-empty';
import {getClassOfferCardName} from '../../utils';
import {OfferCardNames} from '../../constants';

class FavoritesList extends PureComponent {
  constructor(props) {
    super(props);

    this._changeFavoriteOfferClickHandler = this._changeFavoriteOfferClickHandler.bind(this);
  }

  render() {
    const {favoriteOffers} = this.props;

    if (!favoriteOffers.length) {
      return <FavoritesEmpty />;
    }

    const favoriteCities = [...new Set(favoriteOffers.map((offer) => offer.city.name))];
    const classCard = getClassOfferCardName(OfferCardNames.FAVORITE_OFFER);

    return <PageLayout pageName="favorites">
      <Fragment>
        <main className="page__main page__main--favorites">
          <div className="page__favorites-container container">
            <section className="favorites">
              <h1 className="favorites__title">Saved listing</h1>
              <ul className="favorites__list">
                {favoriteCities.map((city, index) => (
                  <li className="favorites__locations-items" key={`${city}-${index}`}>
                    <div className="favorites__locations locations locations--current">
                      <div className="locations__item">
                        <a className="locations__item-link" href="#">
                          <span>{city}</span>
                        </a>
                      </div>
                    </div>
                    <div className="favorites__places">
                      {favoriteOffers.filter((offer) => offer.city.name === city).map((offer) => <OfferCard
                        offerId={offer.id}
                        offer={offer}
                        key={offer.id}
                        activeOfferMouseEnterHandler={() => {}}
                        deactiveOfferMouseLeaveHandler={() => {}}
                        changeFavoriteOfferClickHandler={this._changeFavoriteOfferClickHandler}
                        classCard={classCard}
                        isFavoriteOffer
                      />)}
                    </div>
                  </li>
                ))}
              </ul>
            </section>
          </div>
        </main>
        <footer className="footer container">
          <Link to="/" className="footer__logo-link">
            <img className="footer__logo" src="img/logo.svg" alt="6 cities logo" width="64" height="33" />
          </Link>
        </footer>
      </Fragment>
    </PageLayout>;
  }

  componentDidMount() {
    const {loadFavorites} = this.props;
    loadFavorites();
  }

  _changeFavoriteOfferClickHandler(offerId, isFavorite) {
    const {changeOfferFavorite} = this.props;
    const status = isFavorite === true ? 0 : 1;
    changeOfferFavorite(offerId, status);
  }
}

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  favoriteOffers: state.appReducer.favorites,
});

const mapDispatchToProps = (dispatch) => ({
  loadFavorites: () => dispatch(Operation.loadFavorites()),
  changeOfferFavorite: (offerId, status) => dispatch(Operation.changeOfferFavorite(offerId, status)),
});

FavoritesList.propTypes = {
  favoriteOffers: PropTypes.array.isRequired,
  loadFavorites: PropTypes.func.isRequired,
  changeOfferFavorite: PropTypes.func.isRequired,
};

export {FavoritesList};

export default connect(mapStateToProps, mapDispatchToProps)(FavoritesList);
