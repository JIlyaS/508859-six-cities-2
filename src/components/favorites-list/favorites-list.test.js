import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import {shape} from 'prop-types';
import {BrowserRouter} from 'react-router-dom';
import {MOCK_DATA_UPDATED_FAVORITE} from '../../constants';

import {FavoritesList} from './favorites-list';

Enzyme.configure({adapter: new Adapter()});

const router = {
  history: new BrowserRouter().history,
  route: {
    location: {},
    match: {},
  },
};

const createContext = () => ({
  context: {router},
  childContextTypes: {router: shape({})},
});

it(`FavoritesList correctly renders after relaunch`, () => {
  const tree = shallow(<FavoritesList
    favoriteOffers={MOCK_DATA_UPDATED_FAVORITE}
    loadFavorites={() => {}}
    changeOfferFavorite={() => {}}
  />, createContext());

  expect(toJSON(tree)).toMatchSnapshot();
});
