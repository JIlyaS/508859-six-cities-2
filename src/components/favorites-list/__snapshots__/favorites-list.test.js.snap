// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesList correctly renders after relaunch 1`] = `
<PageLayout
  pageName="favorites"
>
  <main
    className="page__main page__main--favorites"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites"
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <ul
          className="favorites__list"
        >
          <li
            className="favorites__locations-items"
            key="Amsterdam-0"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    Amsterdam
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <OfferCard
                classCard={
                  Object {
                    "card": "favorites__",
                    "info": "favorites__card-info",
                    "wrapper": "favorites",
                  }
                }
                isFavoriteOffer={true}
                key="1"
                offer={
                  Object {
                    "city": Object {
                      "location": Object {
                        "coordinateCity": Array [
                          52.370216,
                          4.895168,
                        ],
                        "zoomCity": 10,
                      },
                      "name": "Amsterdam",
                    },
                    "features": Array [
                      "Entire place",
                      "3 Bedrooms",
                      "Max 4 adults",
                    ],
                    "hostUser": Object {
                      "avatar": "img/1.png",
                      "id": 3,
                      "name": "Angelina",
                      "status": true,
                    },
                    "id": 1,
                    "img": "img/1.png",
                    "insideProperties": Array [
                      "Heating",
                      "Kitchen",
                      "Cable TV",
                      "Washing machine",
                      "Coffee machine",
                      "Dishwasher",
                    ],
                    "isFavorite": true,
                    "isPremium": false,
                    "location": Object {
                      "coordinate": Array [
                        52.35514938496378,
                        4.673877537499948,
                      ],
                      "zoom": 8,
                    },
                    "photos": Array [],
                    "price": 120,
                    "rating": 5,
                    "title": "Beautiful & luxurious studio at great location",
                    "type": "apartment",
                  }
                }
                offerId={1}
                onActiveOfferMouseEnter={[Function]}
                onDeactiveOfferMouseLeave={[Function]}
                onFavoriteOfferClick={[Function]}
              />
            </div>
          </li>
        </ul>
      </section>
    </div>
  </main>
  <footer
    className="footer container"
  >
    <Link
      className="footer__logo-link"
      to="/"
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </Link>
  </footer>
</PageLayout>
`;
